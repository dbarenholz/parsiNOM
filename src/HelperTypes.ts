import { Parser, STypeBase } from './Parser';

export type TupleToUnion<T extends readonly unknown[]> = T[number];

export type DeParser<T extends Parser<unknown>> = T extends Parser<infer U> ? U : never;

export type DeParserArray<T extends readonly Parser<unknown>[]> = {
	[I in keyof T]: T[I] extends Parser<infer R> ? R : never;
};
