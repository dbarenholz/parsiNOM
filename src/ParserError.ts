import { ParseFailure } from './HelperTypes';

function generateMessage(str: string, parseFailure: ParseFailure, verbose: boolean): string {
	const expectedMessage = `Expected ${parseFailure.expected.join(' or ')}`;
	let message = `Parse Failure: ${expectedMessage} at index ${parseFailure.furthest.index}, line ${parseFailure.furthest.line}, column ${parseFailure.furthest.column}.`;

	if (verbose) {
		const lines = str.split('\n');
		const failedLine = lines[parseFailure.furthest.line - 1]; // line is a one based index

		const linePrefix = `${parseFailure.furthest.line} |   `;
		message += `\n\n${linePrefix}${failedLine}`;
		message += `\n${' '.repeat(parseFailure.furthest.column - 1 + linePrefix.length)}^ (${expectedMessage})`;
	}

	return message;
}

export class ParsingError extends Error {
	constructor(str: string, parseFailure: ParseFailure) {
		super(generateMessage(str, parseFailure, true));
	}
}
