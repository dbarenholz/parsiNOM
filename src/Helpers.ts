import { ParseResult, STypeBase } from './HelperTypes';
import { Parser } from './Parser';

export function arrayUnion(a: string[] | never[], b: string[] | never[]): string[] {
	const ret: string[] = [...a];
	for (const bElement of b) {
		let alreadyIncluded = false;
		for (const retElement of ret) {
			if (bElement === retElement) {
				alreadyIncluded = true;
			}
		}
		if (!alreadyIncluded) {
			ret.push(bElement);
		}
	}
	ret.sort();
	return ret;
}

// Parser Helpers

export class ParserHelpers {
	followedBy<SType extends STypeBase>(x: Parser<SType>): Parser<SType> {
		return new Parser<SType>(context => {
			const result = x.p(context.copy());
			result.position = context.position;
			return result;
		});
	}

	/**
	 * Returns a parser that yields undefined if the input parser fails, and fails if the input parser accepts. Consumes no input.
	 * More or less an inverse lookahead.
	 *
	 * @param parser
	 */
	notFollowedBy(parser: Parser<unknown>): Parser<undefined> {
		return new Parser((context): ParseResult<undefined> => {
			const contextCopy = context.copy();
			const result = parser.p(contextCopy);
			const text = context.sliceTo(contextCopy.position.index);
			return result.success ? context.fail(`not '` + text + `'`) : context.succeed(undefined);
		});
	}

	/**
	 * Returns a parser that passes the next character into `fn` and yields the character if `fn` returns true.
	 *
	 * @param fn
	 */
	test(fn: (char: string) => boolean): Parser<string> {
		return new Parser<string>((context): ParseResult<string> => {
			const char = context.input[context.position.index];
			if (!context.atEOF() && fn(char)) {
				return context.succeedOffset(1, char);
			} else {
				return context.fail(`a character matching ${fn}`);
			}
		});
	}
}

export const P_HELPERS = new ParserHelpers();
