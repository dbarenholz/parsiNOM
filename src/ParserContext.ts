import { arrayUnion } from './Helpers';
import { ParseResult, ParsingPosition, STypeBase } from './HelperTypes';

export class ParserContext {
	readonly input: string;
	position: ParsingPosition;

	constructor(input: string, position: ParsingPosition) {
		this.input = input;
		this.position = position;
	}

	moveToPosition(position: ParsingPosition): ParserContext {
		this.position = position;
		return this;
	}

	copy(): ParserContext {
		return new ParserContext(this.input, {
			index: this.position.index,
			column: this.position.column,
			line: this.position.line,
		});
	}

	atEOF(): boolean {
		return this.position.index >= this.input.length;
	}

	// OPTIMIZATION: only create a new position when needed, otherwise mutate
	private advanceTo(index: number): ParsingPosition {
		if (index < this.position.index) {
			throw new Error(`Can not advance backwards. Current pos ${this.position.index}. Advance target index ${index}.`);
		}
		if (index === this.position.index) {
			return this.position;
		}

		let line = this.position.line;
		let column = this.position.column;

		for (let i = this.position.index; i < index; i++) {
			if (this.input[i] === '\n') {
				line += 1;
				column = 1;
			} else {
				column += 1;
			}
		}

		this.position = {
			index: index,
			line: line,
			column: column,
		};

		return this.position;
	}

	private invalidPosition(): ParsingPosition {
		return {
			index: -1,
			line: -1,
			column: -1,
		};
	}

	sliceTo(endIndex: number): string {
		return this.input.slice(this.position.index, endIndex);
	}

	succeedOffset<SType extends STypeBase>(offset: number, value: SType): ParseResult<SType> {
		return this.succeedAt(this.position.index + offset, value);
	}

	failOffset<SType extends STypeBase>(offset: number, expected: string | string[]): ParseResult<SType> {
		return this.failAt(this.position.index + offset, expected);
	}

	succeed<SType extends STypeBase>(value: SType): ParseResult<SType> {
		return this.succeedAt(this.position.index, value);
	}

	fail<SType extends STypeBase>(expected: string | string[]): ParseResult<SType> {
		return this.failAt(this.position.index, expected);
	}

	succeedAt<SType extends STypeBase>(index: number, value: SType): ParseResult<SType> {
		return {
			success: true,
			position: this.advanceTo(index),
			value: value,
			furthest: this.invalidPosition(),
			expected: [],
		};
	}

	failAt<SType extends STypeBase>(index: number, expected: string | string[]): ParseResult<SType> {
		return {
			success: false,
			position: this.invalidPosition(),
			value: null,
			furthest: this.advanceTo(index),
			expected: Array.isArray(expected) ? expected : [expected],
		};
	}

	merge<ASType extends STypeBase, BSType extends STypeBase>(a: ParseResult<ASType> | undefined, b: ParseResult<BSType>): ParseResult<BSType> {
		if (a === undefined) {
			return b;
		}

		if (b.furthest.index > a.furthest.index) {
			return b;
		}

		const expected: string[] = b.furthest.index === a.furthest.index ? arrayUnion(a.expected, b.expected) : a.expected;

		b.furthest = a.furthest;
		b.expected = expected;
		return b;
	}
}
