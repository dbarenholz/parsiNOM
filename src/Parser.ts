import { ParserContext } from './ParserContext';
import { ParseFailure, ParseFunction, ParseResult, ParsingMarker, ParsingNode, STypeBase } from './HelperTypes';
import { P } from './ParsiNOM';
import { P_HELPERS, validateRange } from './Helpers';
import { P_UTILS } from './ParserUtils';

export class Parser<const SType extends STypeBase> {
	public p: ParseFunction<SType>;

	constructor(p: ParseFunction<SType>) {
		this.p = p;
	}

	tryParse(str: string): ParseResult<SType> {
		return this.p(new ParserContext(str, { index: 0, line: 1, column: 1 }));
	}

	parse(str: string): SType | undefined {
		const result = this.tryParse(str);
		if (result.success) {
			return result.value;
		} else {
			throw P.createError(str, result);
		}
	}

	or<OtherSType extends STypeBase>(other: Parser<OtherSType>): Parser<SType | OtherSType> {
		return P.or(this as Parser<SType>, other);
	}

	trim(parser: Parser<unknown>): Parser<SType> {
		return this.wrap(parser, parser);
	}

	wrap(leftParser: Parser<unknown>, rightParser: Parser<unknown>): Parser<SType> {
		return P.sequenceMap(
			(_l, m: SType, _r) => {
				return m;
			},
			leftParser,
			this as Parser<SType>,
			rightParser,
		);
	}

	then<OtherSType extends STypeBase>(next: Parser<OtherSType>): Parser<OtherSType> {
		return P.sequence(this as Parser<SType>, next).map<OtherSType>(results => results[1] as OtherSType);
	}

	many(): Parser<SType[]> {
		return new Parser<SType[]>((context): ParseResult<SType[]> => {
			let result = undefined;
			const startIndex = context.position.index;
			const value: SType[] = [];

			while (true) {
				const contextCopy = context.copy();
				const newResult = this.p(contextCopy);

				result = context.merge(result, newResult);

				if (result.success) {
					if (result.position.index === startIndex) {
						throw new Error('infinite loop in many() parser detected');
					}

					context.moveToPosition(contextCopy.position);
					value.push(result.value);
				} else {
					return context.merge(result, context.succeed(value));
				}
			}
		});
	}

	repeat(min: number, max: number): Parser<SType[]> {
		validateRange(min, max);

		return new Parser<SType[]>((context): ParseResult<SType[]> => {
			let newResult = undefined;
			let result = undefined;
			const value: SType[] = [];
			let iteration = 0;

			for (; iteration < min; iteration++) {
				newResult = this.p(context);

				result = context.merge(result, newResult);

				if (newResult.success) {
					value.push(newResult.value);
				} else {
					return result as ParseFailure;
				}
			}

			for (; iteration < max; iteration++) {
				const contextCopy = context.copy();
				newResult = this.p(contextCopy);

				result = context.merge(result, newResult);

				if (newResult.success) {
					context.moveToPosition(contextCopy.position);
					value.push(newResult.value);
				} else {
					break;
				}
			}

			// first move the result to the context position and then change its value.
			return context.merge(result, context.succeed(value));
		});
	}

	result<OtherSType extends STypeBase>(value: OtherSType): Parser<OtherSType> {
		return this.map(() => value);
	}

	atMost(max: number): Parser<SType[]> {
		return this.repeat(0, max);
	}

	atLeast(min: number): Parser<SType[]> {
		return P.sequenceMap(
			(part1, part2) => {
				return part1.concat(part2);
			},
			this.repeat(min, min),
			this.many(),
		);
	}

	map<const OtherSType extends STypeBase>(fn: (value: SType) => OtherSType): Parser<OtherSType> {
		return new Parser(context => {
			const result = this.p(context);
			if (!result.success) {
				return result;
			}
			// we are kind of changing the generic type of `ParseResult` here and TS does not like it.
			// return context.mutateResult(result, fn(result.value));

			return context.merge(result, context.succeed(fn(result.value)));
		});
	}

	skip<const OtherSType extends STypeBase>(next: Parser<OtherSType>): Parser<SType> {
		return P.sequence(this as Parser<SType>, next).map(value => value[0]);
	}

	mark(): Parser<ParsingMarker<SType>> {
		return P.sequenceMap(
			(start, value: SType, end) => {
				return {
					start: start,
					value: value,
					end: end,
				};
			},
			P_UTILS.position(),
			this as Parser<SType>,
			P_UTILS.position(),
		);
	}

	node(name: string): Parser<ParsingNode<SType>> {
		return P.sequenceMap(
			(start, value: SType, end) => {
				return {
					name: name,
					start: start,
					value: value,
					end: end,
				};
			},
			P_UTILS.position(),
			this as Parser<SType>,
			P_UTILS.position(),
		);
	}

	separateBy(separator: Parser<unknown>): Parser<SType[]> {
		return P.separateBy(this, separator);
	}

	separateByNotEmpty(separator: Parser<unknown>): Parser<SType[]> {
		return P.separateByNotEmpty(this, separator);
	}

	/**
	 * Functions like lookahead. Checks if this parser is followed by `next`, but does not advance the parsing position.
	 *
	 * @param next
	 */
	followedBy(next: Parser<unknown>): Parser<SType> {
		return this.skip(P_HELPERS.followedBy(next));
	}

	/**
	 * Functions like inverse lookahead. Checks if this parser is **not** followed by `next`, but does not advance the parsing position.
	 *
	 * @param next
	 */
	notFollowedBy(next: Parser<unknown>): Parser<SType> {
		return this.skip(P_HELPERS.notFollowedBy(next));
	}

	describe(expected: string | string[] = []): Parser<SType> {
		if (!Array.isArray(expected)) {
			expected = [expected];
		}

		return new Parser<SType>(context => {
			const result = this.p(context);
			if (!result.success) {
				result.expected = expected as string[];
			}
			return result;
		});
	}

	/**
	 * Makes the parser optional. Provide an optional fallback value that will be yielded when the parser fails, otherwise the parser yields `undefined`, but it will never fail.
	 */
	optional(): Parser<SType | undefined>;
	optional<OtherSType extends STypeBase>(value: OtherSType): Parser<SType | OtherSType>;
	optional<OtherSType extends STypeBase>(value?: OtherSType): Parser<SType | OtherSType | undefined> {
		return this.or(P.succeed(value));
	}

	/**
	 * Specify a function that returns a followup parser based on the result of this parser.
	 *
	 * @param fn
	 */
	chain<OtherSType extends STypeBase>(fn: (result: SType) => Parser<OtherSType>): Parser<OtherSType> {
		return new Parser<OtherSType>(context => {
			const result: ParseResult<SType> = this.p(context);
			if (!result.success) {
				return result;
			}
			const nextParser: Parser<OtherSType> = fn(result.value);
			const nextResult = nextParser.p(context);
			return context.merge(result, nextResult);
		});
	}

	thenEof(): Parser<SType> {
		return new Parser<SType>(context => {
			const result: ParseResult<SType> = this.p(context);
			if (!result.success) {
				return result;
			}
			if (!context.atEOF()) {
				return context.merge(result, context.fail('eof'));
			}
			return result;
		});
	}
}
