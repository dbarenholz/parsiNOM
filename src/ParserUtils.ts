import { Parser } from './Parser';
import { P } from './ParsiNOM';
import { ParsingPosition, ParsingRange } from './HelperTypes';

export class P_UTILS {
	/**
	 * Yields the current position.
	 */
	static position(): Parser<ParsingPosition> {
		return new Parser<ParsingPosition>(context => {
			return context.succeed(context.getPosition());
		});
	}

	/**
	 * Accepts any single character except for eof.
	 * Yields the character
	 */
	static any(): Parser<string> {
		return new Parser<string>(context => {
			if (context.atEOF()) {
				return context.fail('any character');
			}
			return context.succeedOffset(1, context.input[context.position.index]);
		});
	}

	/**
	 * Accepts the entire rest of the string until the end.
	 * Yields the rest of the string.
	 */
	static remaining(): Parser<string> {
		return new Parser<string>(context => {
			return context.succeedAt(context.input.length, context.input.slice(context.position.index));
		});
	}

	static eof(): Parser<undefined> {
		return new Parser<undefined>(context => {
			if (!context.atEOF()) {
				return context.fail('eof');
			}
			return context.succeed(undefined);
		});
	}

	static digit() {
		return P.regexp(/^[0-9]/).describe('a digit');
	}

	static digits() {
		return P.regexp(/^[0-9]+/).describe('optional digits');
	}

	static letter() {
		return P.regexp(/^[a-z]/i).describe('a letter');
	}

	static letters() {
		return P.regexp(/^[a-z]+/i).describe('optional letters');
	}

	static optionalWhitespace() {
		return P.regexp(/^\s*/).describe('optional whitespace');
	}

	static whitespace() {
		return P.regexp(/^\s+/).describe('whitespace');
	}

	static cr() {
		return P.string('\r');
	}

	static lf() {
		return P.string('\n');
	}

	static crlf() {
		return P.string('\r\n');
	}

	static newline() {
		return P.or(this.crlf(), this.lf(), this.cr()).describe('newline');
	}

	static prefix<OperatorSType, OtherSType, ReturnSType>(
		operatorsParser: Parser<OperatorSType>,
		nextParser: Parser<OtherSType>,
		combine: (a: OperatorSType, b: OtherSType | ReturnSType) => ReturnSType,
	): Parser<OtherSType | ReturnSType> {
		// const parser: Parser<OtherSType | ReturnSType> = P.reference(() => {
		// 	return P.sequenceMap(combine, operatorsParser, parser).or(nextParser);
		// });
		// return parser;

		return P.sequenceMap(
			(prefixes, x) => {
				return prefixes.reduce<OtherSType | ReturnSType>((acc, y) => combine(y, acc), x);
			},
			operatorsParser.many(),
			nextParser,
		);
	}

	static postfix<OperatorSType, OtherSType, ReturnSType>(
		operatorsParser: Parser<OperatorSType>,
		nextParser: Parser<OtherSType>,
		combine: (a: OperatorSType, b: OtherSType | ReturnSType) => ReturnSType,
	): Parser<OtherSType | ReturnSType> {
		return P.sequenceMap(
			(x, postfixes) => {
				return postfixes.reduce<OtherSType | ReturnSType>((acc, y) => combine(y, acc), x);
			},
			nextParser,
			operatorsParser.many(),
		);
	}

	static binaryRight<OperatorSType, OtherSType, ReturnSType>(
		operatorsParser: Parser<OperatorSType>,
		operandParser: Parser<OtherSType>,
		combine: (a: OtherSType, b: OperatorSType, c: OtherSType | ReturnSType) => ReturnSType,
	): Parser<OtherSType | ReturnSType> {
		return P.sequenceMap(
			(others, last: OtherSType) => {
				return others.reverse().reduce<OtherSType | ReturnSType>((acc, y) => {
					const [operand, operator] = y;
					return combine(operand, operator, acc);
				}, last);
			},
			P.sequence(operandParser, operatorsParser.trim(this.optionalWhitespace())).many(),
			operandParser,
		);
	}

	static binaryLeft<OperatorSType, OtherSType, ReturnSType>(
		operatorsParser: Parser<OperatorSType>,
		operandParser: Parser<OtherSType>,
		combine: (a: OtherSType | ReturnSType, b: OperatorSType, c: OtherSType) => ReturnSType,
	): Parser<OtherSType | ReturnSType> {
		return P.sequenceMap(
			(first: OtherSType, others) => {
				return others.reduce<OtherSType | ReturnSType>((acc, y) => {
					const [operator, operand] = y;
					return combine(acc, operator, operand);
				}, first);
			},
			operandParser,
			P.sequence(operatorsParser.trim(this.optionalWhitespace()), operandParser).many(),
		);
	}

	static binaryRightRange<OperatorSType, OtherSType, ReturnSType>(
		operatorsParser: Parser<OperatorSType>,
		operandParser: Parser<OtherSType>,
		combine: (range: ParsingRange, a: OtherSType, b: OperatorSType, c: OtherSType | ReturnSType) => ReturnSType,
	): Parser<OtherSType | ReturnSType> {
		return P.sequenceMap(
			(others, last: OtherSType, to) => {
				return others.reverse().reduce<OtherSType | ReturnSType>((acc, y) => {
					const [from, operand, operator] = y;
					return combine({ from, to }, operand, operator, acc);
				}, last);
			},
			P.sequence(P_UTILS.position(), operandParser, operatorsParser.trim(this.optionalWhitespace())).many(),
			operandParser,
			P_UTILS.position(),
		);
	}

	static binaryLeftRange<OperatorSType, OtherSType, ReturnSType>(
		operatorsParser: Parser<OperatorSType>,
		operandParser: Parser<OtherSType>,
		combine: (range: ParsingRange, a: OtherSType | ReturnSType, b: OperatorSType, c: OtherSType) => ReturnSType,
	): Parser<OtherSType | ReturnSType> {
		return P.sequenceMap(
			(from, first: OtherSType, others) => {
				return others.reduce<OtherSType | ReturnSType>((acc, y) => {
					const [operator, operand, to] = y;
					return combine({ from, to }, acc, operator, operand);
				}, first);
			},
			P_UTILS.position(),
			operandParser,
			P.sequence(operatorsParser.trim(this.optionalWhitespace()), operandParser, P_UTILS.position()).many(),
		);
	}

	static func<ArgsSType, ReturnSType>(
		name: string | Parser<string>,
		args: Parser<ArgsSType>,
		combine: (name: string, args: ArgsSType) => ReturnSType,
	): Parser<ReturnSType> {
		const nameParser = typeof name === 'string' ? P.string(name) : name;
		return P.sequenceMap(
			(name, _lBracket, _1, args, _2, _rBracket) => {
				return combine(name, args);
			},
			nameParser,
			P.string('('),
			this.optionalWhitespace(),
			args,
			this.optionalWhitespace(),
			P.string(')'),
		);
	}
}
