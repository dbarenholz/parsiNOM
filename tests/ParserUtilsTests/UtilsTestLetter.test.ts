import { P_UTILS } from '../../src/ParserUtils';
import { testParser } from '../TestHelpers';
describe.each([
	['a', true],
	['b', true],
	['z', true],
	['#', false],
	[' ', false],
	['', false],
])(`letter`, (str, shouldSucceed) => {
	const parser = P_UTILS.letter();
	testParser(parser, str, shouldSucceed);
});

describe.each([
	['a', true],
	['bb', true],
	['zz', true],
	['#', false],
	[' ', false],
	['', false],
])(`letters`, (str, shouldSucceed) => {
	const parser = P_UTILS.letters();
	testParser(parser, str, shouldSucceed);
});
