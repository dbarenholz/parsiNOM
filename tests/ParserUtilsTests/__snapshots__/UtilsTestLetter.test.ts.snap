// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`letter 'a' conditional success tests AST matches snapshot 1`] = `"a"`;

exports[`letter 'b' conditional success tests AST matches snapshot 1`] = `"b"`;

exports[`letter 'z' conditional success tests AST matches snapshot 1`] = `"z"`;

exports[`letter '#' conditional failure tests Error matches snapshot 1`] = `
{
  "expected": [
    "a letter",
  ],
  "pos": {
    "column": 1,
    "index": 0,
    "line": 1,
  },
}
`;

exports[`letter ' ' conditional failure tests Error matches snapshot 1`] = `
{
  "expected": [
    "a letter",
  ],
  "pos": {
    "column": 1,
    "index": 0,
    "line": 1,
  },
}
`;

exports[`letter '' conditional failure tests Error matches snapshot 1`] = `
{
  "expected": [
    "a letter",
  ],
  "pos": {
    "column": 1,
    "index": 0,
    "line": 1,
  },
}
`;

exports[`letters 'a' conditional success tests AST matches snapshot 1`] = `"a"`;

exports[`letters 'bb' conditional success tests AST matches snapshot 1`] = `"bb"`;

exports[`letters 'zz' conditional success tests AST matches snapshot 1`] = `"zz"`;

exports[`letters '#' conditional failure tests Error matches snapshot 1`] = `
{
  "expected": [
    "optional letters",
  ],
  "pos": {
    "column": 1,
    "index": 0,
    "line": 1,
  },
}
`;

exports[`letters ' ' conditional failure tests Error matches snapshot 1`] = `
{
  "expected": [
    "optional letters",
  ],
  "pos": {
    "column": 1,
    "index": 0,
    "line": 1,
  },
}
`;

exports[`letters '' conditional failure tests Error matches snapshot 1`] = `
{
  "expected": [
    "optional letters",
  ],
  "pos": {
    "column": 1,
    "index": 0,
    "line": 1,
  },
}
`;
